wandb/util.py

```diff
@@ -146,22 +168,75 @@ def sentry_reraise(exc: Any) -> None:
 
 
 def sentry_set_scope(
-    process_context: Optional[str],
-    entity: Optional[str],
-    project: Optional[str],
-    email: Optional[str] = None,
-    url: Optional[str] = None,
+    settings_dict: Optional[
+        Union[
+            "wandb.sdk.wandb_settings.Settings",
+            "wandb.sdk.internal.settings_static.SettingsStatic",
+        ]
+    ] = None,
+    process_context: Optional[str] = None,
 ) -> None:
     # Using GLOBAL_HUB means these tags will persist between threads.
     # Normally there is one hub per thread.
+
+    # Tags come from two places: settings and args passed into this func.
+    args = dict(locals())
+    del args["settings_dict"]
+
+    settings_tags = [
+        "entity",
+        "project",
+        "run_id",
+        "run_url",
+        "sweep_url",
+        "sweep_id",
+        "deployment",
+        "_require_service",
+    ]
+
+    s = settings_dict
+
+    # convenience function for getting attr from settings
+    def get(key: str) -> Any:
+        return getattr(s, key, None)
+
     with sentry_sdk.hub.GLOBAL_HUB.configure_scope() as scope:
-        scope.set_tag("process_context", process_context)
-        scope.set_tag("entity", entity)
-        scope.set_tag("project", project)
-        if email:
-            scope.user = {"email": email}
-        if url:
-            scope.set_tag("url", url)
+        scope.set_tag("platform", get_platform_name())
+
+        # apply settings tags
+        if s is not None:
+            for tag in settings_tags:
+                val = get(tag)
+                if val not in [None, ""]:
+                    scope.set_tag(tag, val)
+
+            python_runtime = (
+                "colab"
+                if get("_colab")
+                else ("jupyter" if get("_jupyter") else "python")
+            )
+            scope.set_tag("python_runtime", python_runtime)
+
+            # Hack for constructing run_url and sweep_url given run_id and sweep_id
+            required = ["entity", "project", "base_url"]
+            params = {key: get(key) for key in required}
+            if all(params.values()):
+                # here we're guaranteed that entity, project, base_url all have valid values
+                app_url = wandb.util.app_url(params["base_url"])
+                e, p = [quote(params[k]) for k in ["entity", "project"]]
+
+                for word in ["run", "sweep"]:
+                    _url, _id = f"{word}_url", f"{word}_id"
+                    if not get(_url) and get(_id):
+                        scope.set_tag(_url, f"{app_url}/{e}/{p}/{word}s/{get(_id)}")
```

Comment:
Maybe add a TODO here? Also I still don't understand how you have entity and project and don't have the urls...

Reply:
@kptkin 
we are going to clean this up after release.  We know what needs to be done, just wanted to push it off for now.    It is what we talked about, passing the info back to the internal process.

Reply:
ok... sounds good. Let's leave a TODO here :)

Reply:
```suggestion
            if all(params.values()):
                # here we're guaranteed that entity, project, base_url all have valid values
                app_url = wandb.util.app_url(params["base_url"])
                e, p = [quote(params[k]) for k in ["entity", "project"]]

                # TODO: the settings object will be updated to contain run_url and sweep_url
                # This is done by passing a settings_map in the run_start protocol buffer message
                for word in ["run", "sweep"]:
                    _url, _id = f"{word}_url", f"{word}_id"
                    if not get(_url) and get(_id):
                        scope.set_tag(_url, f"{app_url}/{e}/{p}/{word}s/{get(_id)}")
```

Reply:
done